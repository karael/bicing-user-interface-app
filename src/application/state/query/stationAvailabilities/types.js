import { defineAction } from 'redux-define';

import { START, PENDING, CANCELLED, SUCCESS, FAILURE } from 'application/state/stateConstants';

const domain = defineAction('stationAvailabilities');

export const FETCH = defineAction(
    'FETCH',
    [START, PENDING, CANCELLED, SUCCESS, FAILURE],
    domain
);
