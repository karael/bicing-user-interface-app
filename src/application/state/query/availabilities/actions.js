import { FETCH } from 'application/state/query/availabilities/types';

export const fetchStart = (
  itineraryStep,
  itineraryAt,
  periodStartAt,
  periodEndAt,
  interval,
  ...stationIds
) => ({
  error: false,
  meta: { isFetching: true, itineraryStep, itineraryAt },
  payload: {
    periodStartAt, periodEndAt, interval, stationIds,
  },
  type: FETCH.START,
});
export const fetchPending = itineraryStep => ({
  error: false,
  meta: { isFetching: true, itineraryStep },
  type: FETCH.PENDING,
});
export const fetchSuccess = (itineraryStep, data) => ({
  error: false,
  meta: { isFetching: false, itineraryStep },
  type: FETCH.SUCCESS,
  payload: data,
});
export const fetchFailure = (itineraryStep, error) => ({
  error: true,
  meta: { isFetching: false, itineraryStep },
  type: FETCH.FAILURE,
  payload: error,
});
