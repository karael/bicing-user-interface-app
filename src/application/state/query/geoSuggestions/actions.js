import { FETCH } from 'application/state/query/geoSuggestions/types';

export const fetchStart = () => ({
  error: false,
  meta: { isFetching: false },
  type: FETCH.START,
});
export const fetchPending = () => ({
  error: false,
  meta: { isFetching: true },
  type: FETCH.PENDING,
});
export const fetchSuccess = geoSuggestions => ({
  error: false,
  meta: { isFetching: false },
  type: FETCH.SUCCESS,
  payload: geoSuggestions,
});
export const fetchFailure = error => ({
  error: true,
  meta: { isFetching: false },
  type: FETCH.FAILURE,
  payload: error,
});
