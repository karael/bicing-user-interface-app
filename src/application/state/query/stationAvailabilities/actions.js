import { FETCH } from 'application/state/query/stationAvailabilities/types';

export const fetchStart = (stationId, periodStart, periodEnd, interval) => ({
  error: false,
  meta: { isFetching: true },
  payload: {
    stationId, periodStart, periodEnd, interval,
  },
  type: FETCH.START,
});
export const fetchPending = () => ({
  error: false,
  meta: { isFetching: true },
  type: FETCH.PENDING,
});
export const fetchSuccess = data => ({
  error: false,
  meta: { isFetching: false },
  type: FETCH.SUCCESS,
  payload: data,
});
export const fetchFailure = error => ({
  error: true,
  meta: { isFetching: false },
  type: FETCH.FAILURE,
  payload: error,
});
