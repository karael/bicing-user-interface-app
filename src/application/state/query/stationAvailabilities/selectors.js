export const isFetchingSelector = (state) => state.query.stationAvailabilities.payload.isFetching;
export const dataSelector = (state) => state.query.stationAvailabilities.data;
export const filterSelector = (state) => {
    const data = dataSelector(state);

    return data && data['filter'] ? data['filter'] : null;
};
export const stationAvailabilitiesSelector = (state) => {
    const data = dataSelector(state);

    return data && data['availabilities'] ? data['availabilities'] : null;
};
export const stationAvailabilitiesSortedSelector = (state) => {
    const availabilities = stationAvailabilitiesSelector(state);

    if (availabilities === null) {
        return null;
    }

    const datasetslabels = [];
    const datasetsAvailableSlotMin = [], datasetsAvailableSlotMax = [], datasetsAvailableSlotAvg = [];
    const datasetsAvailableBikeMin = [], datasetsAvailableBikeMax = [], datasetsAvailableBikeAvg = [];

    availabilities.map(availability => {
        datasetslabels.push(availability['interval']);
        datasetsAvailableSlotMin.push(availability['available_slot_min']);
        datasetsAvailableSlotMax.push(availability['available_slot_max']);
        datasetsAvailableSlotAvg.push(availability['available_slot_avg']);
        datasetsAvailableBikeMin.push(availability['available_bike_min']);
        datasetsAvailableBikeMax.push(availability['available_bike_max']);
        datasetsAvailableBikeAvg.push(availability['available_bike_avg']);
    });

    return {
        datasetslabels,
        datasetsAvailableSlotMin,
        datasetsAvailableSlotMax,
        datasetsAvailableSlotAvg,
        datasetsAvailableBikeMin,
        datasetsAvailableBikeMax,
        datasetsAvailableBikeAvg,
    };
};
