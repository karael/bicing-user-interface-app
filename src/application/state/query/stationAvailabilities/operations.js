import ByIntervalInPeriodFilter from 'application/state/filter/ByIntervalInPeriodFilter';
import {
  fetchFailure,
  fetchPending,
  fetchSuccess,
} from 'application/state/query/stationAvailabilities/actions';
import { FETCH } from 'application/state/query/stationAvailabilities/types';
import { stationAvailabilitiesType } from 'domain/types/stationAvailabilitiesType';
import { stationIdType } from 'domain/types/stationType';
import HttpStationAvailabilitiesQuery
  from 'infrastructure/bicingApi/HttpStationAvailabilitiesQuery';
import Joi from 'joi';
import { call, put, takeLatest } from 'redux-saga/effects';

export function* fetch(action) {
  yield put(fetchPending());

  try {
    const {
      stationId, periodStart, periodEnd, interval,
    } = action.payload;

    Joi.assert(stationId, stationIdType);

    const stationAvailabilities = yield call(
      HttpStationAvailabilitiesQuery.find,
      stationId,
      ByIntervalInPeriodFilter.fromRawValues(periodStart, periodEnd, interval),
    );

    Joi.assert(stationAvailabilities, stationAvailabilitiesType);

    yield put(fetchSuccess(stationAvailabilities));
  } catch (exception) {
    yield put(fetchFailure(exception));
  }
}

export default function* operation() {
  yield takeLatest(FETCH.START, fetch);
}
