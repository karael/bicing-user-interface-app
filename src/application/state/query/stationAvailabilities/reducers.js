import { FETCH } from 'application/state/query/stationAvailabilities/types';
import produce from 'immer';
import { createReducer } from 'reduxsauce';

export const INITIAL_STATE = {
  data: undefined,
  error: false,
  isFetching: false,
};

export const fetchStart = (state = INITIAL_STATE, action) => produce(state, (draft) => {
  draft.data = undefined;
  draft.error = action.error;
  draft.isFetching = action.meta.isFetching;
});

export const fetchPending = (state = INITIAL_STATE, action) => produce(state, (draft) => {
  draft.data = undefined;
  draft.error = action.error;
  draft.isFetching = action.meta.isFetching;
});

export const fetchSuccess = (state = INITIAL_STATE, action) => produce(state, (draft) => {
  draft.data = action.payload;
  draft.error = action.error;
  draft.isFetching = action.meta.isFetching;
});

export const fetchFailure = (state = INITIAL_STATE, action) => produce(state, (draft) => {
  draft.data = action.payload;
  draft.error = action.error;
  draft.isFetching = action.meta.isFetching;
});

const HANDLERS = {
  [FETCH.START]: fetchStart,
  [FETCH.PENDING]: fetchPending,
  [FETCH.SUCCESS]: fetchSuccess,
  [FETCH.FAILURE]: fetchFailure,
};

export default createReducer(INITIAL_STATE, HANDLERS);
