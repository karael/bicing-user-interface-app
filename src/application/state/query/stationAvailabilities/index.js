import reducer from 'application/state/query/stationAvailabilities/reducers';
import operation from 'application/state/query/stationAvailabilities/operations';
import * as actions from 'application/state/query/stationAvailabilities/actions';
import * as Types from 'application/state/query/stationAvailabilities/types';
import * as selectors from 'application/state/query/stationAvailabilities/selectors';

export {
    actions,
    Types,
    operation,
    selectors
};

export default reducer;
