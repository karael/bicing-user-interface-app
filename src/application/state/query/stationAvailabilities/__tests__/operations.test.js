import * as actions from 'application/state/query/stationAvailabilities/actions';
import operation, { fetch } from 'application/state/query/stationAvailabilities/operations';
import { FETCH } from 'application/state/query/stationAvailabilities/types';
import HttpStationAvailabilityQuery from 'infrastructure/bicingApi/HttpStationAvailabilityQuery';
import { expectSaga, testSaga } from 'redux-saga-test-plan';
import * as matchers from 'redux-saga-test-plan/matchers';
import { throwError } from 'redux-saga-test-plan/providers';
import { v4 as uuid } from 'uuid';

// @todo add byFilter logic
describe('application/state/query/stationAvailabilities/operations', () => {
  test('should wait for a fetch start event to list with operation()', () => {
    testSaga(operation)
      .next()
      .takeLatestEffect(FETCH.START, fetch);
  });

  test('should dispatch a fetchPending action with fetch() generator', () => {
    testSaga(fetch)
      .next()
      .put(actions.fetchPending());
  });

  test('should list expected station with fetch() generator', () => {
    const stationId = uuid();
    const fakeStationAvailabilities = [
      {
        interval: '2018-11-05 14:45:00',
        available_bike_avg: '20.0000000000000000',
        available_bike_min: 20,
        available_bike_max: 20,
        available_slot_avg: '9.0000000000000000',
        available_slot_min: 9,
        available_slot_max: 9,
      },
      {
        interval: '2018-11-05 14:50:00',
        available_bike_avg: '20.0000000000000000',
        available_bike_min: 19,
        available_bike_max: 21,
        available_slot_avg: '9.0000000000000000',
        available_slot_min: 8,
        available_slot_max: 10,
      },
    ];
    const action = { type: FETCH.START, payload: { stationId } };

    return expectSaga(fetch, action)
      .provide([
        [matchers.call.fn(HttpStationAvailabilityQuery.find, stationId), fakeStationAvailabilities],
      ])
      .put(actions.fetchSuccess(fakeStationAvailabilities))
      .run();
  });

  test('should handle error when api call failed in fetch() generator', () => {
    const stationId = uuid();
    const error = new Error('error_api_call');
    const action = { type: FETCH.START, payload: { stationId } };

    return expectSaga(fetch, action)
      .provide([
        [matchers.call.fn(HttpStationAvailabilityQuery.find, stationId), throwError(error)],
      ])
      .put(actions.fetchFailure(error))
      .run();
  });
});
