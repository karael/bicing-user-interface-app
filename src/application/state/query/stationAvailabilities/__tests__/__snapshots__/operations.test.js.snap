// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`application/state/query/stationAvailabilities/operations should handle error when ByIntervalInPeriodFilter is not validated 1`] = `
Object {
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "error": false,
          "meta": Object {
            "isFetching": true,
          },
          "type": "query/stationAvailabilities/FETCH_PENDING",
        },
        "channel": null,
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "error": true,
          "meta": Object {
            "isFetching": false,
          },
          "payload": Object {
            "actual": false,
            "code": "ERR_ASSERTION",
            "expected": true,
            "generatedMessage": false,
            "name": "AssertionError [ERR_ASSERTION]",
            "operator": "==",
          },
          "type": "query/stationAvailabilities/FETCH_FAILURE",
        },
        "channel": null,
      },
    },
  ],
}
`;

exports[`application/state/query/stationAvailabilities/operations should handle error when api call response does not contains expected schema type 1`] = `
Object {
  "call": Array [
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          "a41070e4-3866-490d-a8ab-5195eae71f93",
          Object {
            "interval": "5 min",
            "periodEnd": "2018-08-05 14:51:56",
            "periodStart": "2017-08-05 14:43:56",
          },
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/find",
      },
    },
  ],
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "error": false,
          "meta": Object {
            "isFetching": true,
          },
          "type": "query/stationAvailabilities/FETCH_PENDING",
        },
        "channel": null,
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "error": true,
          "meta": Object {
            "isFetching": false,
          },
          "payload": Object {
            "_object": Array [
              Object {
                "available_bike_avg": "20.0000000000000000",
              },
            ],
            "annotate": "@@redux-saga-test-plan/json/function/<anonymous>",
            "details": Array [
              Object {
                "context": Object {
                  "key": undefined,
                  "label": "value",
                  "value": Array [
                    Object {
                      "available_bike_avg": "20.0000000000000000",
                    },
                  ],
                },
                "message": "\\"value\\" must be an object",
                "path": Array [],
                "type": "object.base",
              },
            ],
            "isJoi": true,
            "name": "ValidationError",
          },
          "type": "query/stationAvailabilities/FETCH_FAILURE",
        },
        "channel": null,
      },
    },
  ],
}
`;

exports[`application/state/query/stationAvailabilities/operations should handle error when stationId is not validated 1`] = `
Object {
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "error": false,
          "meta": Object {
            "isFetching": true,
          },
          "type": "query/stationAvailabilities/FETCH_PENDING",
        },
        "channel": null,
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "error": true,
          "meta": Object {
            "isFetching": false,
          },
          "payload": Object {
            "_object": "not an uuid",
            "annotate": "@@redux-saga-test-plan/json/function/<anonymous>",
            "details": Array [
              Object {
                "context": Object {
                  "key": undefined,
                  "label": "value",
                  "value": "not an uuid",
                },
                "message": "\\"value\\" must be a valid GUID",
                "path": Array [],
                "type": "string.guid",
              },
            ],
            "isJoi": true,
            "name": "ValidationError",
          },
          "type": "query/stationAvailabilities/FETCH_FAILURE",
        },
        "channel": null,
      },
    },
  ],
}
`;
